### ISSUE TITLE
Script with a main method is not run in kscript

### ISSUE THREAD

**Issue by xgouchet** [holgerbrandl/kscript#31]
I'm using IntelliJ to write kotlin script. In order to make my script valid with IntelliJ inspections, I'm defining a standard main method. The problem is that when running my file from `kscript`, my main method is not run at all. 

Is there a special action to run the main method from the command line ?

**Comment by holgerbrandl** [comment]
Not sure if I'm missing your point here, but there is no need for a main method in a kotlin `kts` script. See  https://kotlinlang.org/docs/tutorials/command-line.html#using-the-command-line-to-run-scripts

I could imagine that you're trying to provide a regular kotlin class file (which indeed would need a main method as entry point) to `kscript`. However Kotlin kts scripts allow for more simplistic code. E.g.
```
print(1+1)
```
is a valid kotlin script.

Regular class files are not supported (yet) by `kscript`. It crossed my mind to do so, but it is conceptually not possible without making assumptions about the user code, since a kt file could contain multiple classes with a main methods, so the entry point might not be always clearly defined. 

I hope this answers your question.

Inspections work fine without. If not it's most likely an IJ bug.

I'll update docs to point out the difference between scripts and regular kotlin classes sources more clear.

**Comment by holgerbrandl** [comment]
One potential fix would be via another configuration comment, similar to `//DEPS`. E.g. 
```
//ENTRY Foo.main
``` 
could be implemented to  instruct `kscript` to use the companion method `main` of the class `Foo` as entrypoint. Technically it would do so by adding  the line
```
Foo.main(args)
```
to the end of the provided user code to launch the application. So we would implicitly treat kotlin `kt` files as `kts` files. I think it is legit to do so, but I'm not really sure.

I'm also unsure if such a feature would be conceptually too confusing. Furthermore, such a solution would still require some effort from the user to define `//ENTRY`. 
One the plus side, it would extend `kscript` to also become a more generic application launcher.

Your opinions are welcome here!

**Comment by xgouchet** [comment]
Thanks a lot for all these precisions. 

To explain my issue with IntelliJ, let's use the following script example : 

```
    #!/usr/bin/env kscript
    
    println("Hello from Kotlin!")
    for (arg in args) {
        println("arg: $arg")
    }
```

Even if the file has a .kts extension, IntelliJ complains that the `args` reference is undefined. This means that writing kscript with all the power of IJ is not possible.

**Comment by xgouchet** [comment]
Regarding your //ENTRY directive, this would indeed be a good thing to specify, but you could use the default main function (if any), eg : 

```
#!/usr/bin/env kscript

fun main(args: Array<String>) {
    args.forEach { it -> println(it) }
}
```

**Comment by holgerbrandl** [comment]
Concerning your first comment. This is a known regression In IJ since kotlin v1.1, which is indeed super-annoying. Please vote for it under https://youtrack.jetbrains.com/issue/KT-16802

**Comment by holgerbrandl** [comment]
Concerning the //ENTRY directive. Assuming the presence of a main method does not make sense imho because kotlin scripts don't require a `main` and work fine without. 

And since Kotlin is very flexible concerning how `main` can be defined in a `kt` source file, I'm not sure it's intuitive to assume one particular type of `main`.  E.g. another popular to define via a companion method:
```kotlin
class Foo{
    companion object {
        @JvmStatic fun main(args: Array<String>) {
            println("Hello World")
        }
    }
}
```
To detect the entry-point programmatically here would require quite some non-trivial logic.

**Comment by holgerbrandl** [comment]
I finally managed to implement it for `kt` files (because the only context in which an entry-point declaration makes sense to me.

See [here](https://github.com/holgerbrandl/kscript#does-it-also-work-for-regular-kotlin-class-files-kt) for an example about how to use the new //ENTRY directive.

Please have a look and let me know what you think about it.

**Comment by fjay** [comment]
The main method is still not available with the latest version 1.6.0 

kscript K2.kt
K2.kt:10:20: warning: parameter 'name' is never used
fun findUserByName(name: String): User = null!!
                   ^
.Foor: could not find or load main class test

**Comment by holgerbrandl** [comment]
Thanks for the feedback. Could you run 
```
kscript --clear-cache
kscript --version
```
first to make sure that you're running the latest version and caching is not causing the error.

Btw the user warning is normal, as the argument `name` is indeed not used. I'll update the example.

Do you use the provided example? If not, could you attach your K2.kt to this ticket?

**Comment by fjay** [comment]
I've run 
`kscript --clear-cache`
`kscript --version`

I'm sure it's in v1.6.0,  and it doesn't work for me.

 The content of K2.kt is what you provided [before](https://github.com/holgerbrandl/kscript#ok-but-does-kscript-also-work-for-regular-kotlin-kt-source-files--with-a-main-as-entry-point)

**Comment by holgerbrandl** [comment]
Works for perfectly for me.

What is your kotlin version? Mine is the latest 1.1.4-2 (JRE 1.8.0_131-b11). It also worked with 1.1.3. What is your OS?

Are you sure that this is the correct jar? The layout should/must be:
![image](https://user-images.githubusercontent.com/200952/29763952-9a0397c6-8bd5-11e7-94fe-2d2cab4511a4.png)

**Comment by fjay** [comment]
Here is my runtime environment:
* Kotlin version 1.1.4-2 (JRE 1.8.0_111-b14)
* Ubuntu 14.04.4

An error occurred in this environment: could not find or load main class test

But after I switched to macOS, the same code worked.
I think this feature doesn't work in some operating systems.

**Comment by holgerbrandl** [comment]
It worked for me on ubuntu linux and centos previously, but I'll test the test suite on these systems later today.

**Comment by holgerbrandl** [comment]
fixed with v2.0.1. Feel welcome to comment on this ticket if you still struggle to run your code with that version.
